images/hub/jupyterhub_config.py

```diff
@@ -50,42 +76,106 @@
 
 c.KubeSpawner.namespace = os.environ.get('POD_NAMESPACE', 'default')
 
-
 for trait, cfg_key in (
-    ('start_timeout', 'start-timeout'),
-    ('image_pull_policy', 'image-pull-policy'),
-    ('image_pull_secrets', 'image-pull-secret-name'),
+    ('start_timeout', None),
+    ('image_pull_policy', None),
     ('events_enabled', 'events'),
-    ('extra_labels', 'extra-labels'),
-    ('extra_annotations', 'extra-annotations'),
-    ('uid', 'uid'),
-    ('fs_gid', 'fs-gid'),
-    ('service_account', 'service-account-name'),
-    ('scheduler_name', 'scheduler-name'),
+    ('extra_labels', None),
+    ('extra_annotations', None),
+    ('uid', None),
+    ('fs_gid', None),
+    ('service_account', 'serviceAccountName'),
+    ('storage_extra_labels', 'storage.extraLabels'),
+    ('tolerations', None),
+    ('node_selector', None),
+    ('node_affinity_required', None),
+    ('node_affinity_preferred', None),
+    ('pod_affinity_required', None),
+    ('pod_affinity_preferred', None),
+    ('pod_anti_affinity_required', None),
+    ('pod_anti_affinity_preferred', None),
+    ('lifecycle_hooks', None),
+    ('init_containers', None),
+    ('extra_containers', None),
+    ('mem_limit', 'memory.limit'),
+    ('mem_guarantee', 'memory.guarantee'),
+    ('cpu_limit', 'cpu.limit'),
+    ('cpu_guarantee', 'cpu.guarantee'),
+    ('extra_resource_limits', 'extraResource.limits'),
+    ('extra_resource_guarantees', 'extraResource.guarantees'),
+    ('environment', 'extraEnv'),
 ):
+    if cfg_key is None:
+        cfg_key = camelCaseify(trait)
     set_config_if_not_none(c.KubeSpawner, trait, 'singleuser.' + cfg_key)
-c.KubeSpawner.storage_extra_labels = get_config('singleuser.storage-extra-labels', {})
-
-c.KubeSpawner.image_spec = get_config('singleuser.image-spec')
-c.KubeSpawner.priority_class_name = get_config('singleuser.priority-class-name', "")
-
-c.KubeSpawner.tolerations.extend(get_config('singleuser.tolerations', []))
-c.KubeSpawner.node_selector.update(get_config('singleuser.node-selector', {}))
-c.KubeSpawner.node_affinity_required.extend(get_config('singleuser.node-affinity-required', []))
-c.KubeSpawner.node_affinity_preferred.extend(get_config('singleuser.node-affinity-preferred', []))
-c.KubeSpawner.pod_affinity_required.extend(get_config('singleuser.pod-affinity-required', []))
-c.KubeSpawner.pod_affinity_preferred.extend(get_config('singleuser.pod-affinity-preferred', []))
-c.KubeSpawner.pod_anti_affinity_required.extend(get_config('singleuser.pod-anti-affinity-required', []))
-c.KubeSpawner.pod_anti_affinity_preferred.extend(get_config('singleuser.pod-anti-affinity-preferred', []))
+
+image = get_config("singleuser.image.name")
+if image:
+    tag = get_config("singleuser.image.tag")
+    if tag:
+        image = "{}:{}".format(image, tag)
+
+    c.KubeSpawner.image_spec = image
+
+if get_config('singleuser.imagePullSecret.enabled'):
+    c.KubeSpawner.image_pull_secrets = 'singleuser-image-credentials'
+
+# scheduling:
```

Comment:
Because there is no longer any logic transforming .values before it arrives in the image,
the affinity helpers are no longer used for 'real' users, only the user-placeholder. The logic is the same, though, and much easier to follow/test here in Python.